### ISSUE TITLE
retain always false

### ISSUE THREAD

**Issue by rayrola** [php-mqtt/client#163]
retain always false;
`
MqttClient.php:775
deliverPublishedMessage.
`
retain always false;

**Comment by Namoshek** [comment]
More info required, please provide version details and your test code.

**Comment by reinweb** [comment]
**RETAIN always FALSE  on subscribing Messages.**

I have exactly the same issue in version 1.8.1.
Using the provided example "**/client-examples/02_subscribing/01_subscribe_with_qos_0.php**". 
Line 24ff: 
`$client->subscribe('foo/bar/baz', function (string $topic, string $message, bool $retained) use ($logger, $client) {`
--> the variable $retained is always "false".

Maybe the reason can be found here (as rayrola mentioned on Sep 1) 
`MqttClient.php:775 deliverPublishedMessage.`
and 

``MqttClient.php:760 "false"`

the tool mosquitto_sub shows the correct retain flag:
`mosquitto_sub  -u xxx -P xxx -t 'xxxx#' -k 20 -h xxxx  --verbose  -F '%j' -d --nodelay -q 0 --pretty`

**Comment by Namoshek** [comment]
I'll look into this, may be spec related.

**Comment by Namoshek** [comment]
You are right, it looks like I've never implemented that part for some reason. Implementing it will cause a few breaking changes from an OOP perspective though, so I'll need to see what I can do.

**Comment by Namoshek** [comment]
I've put something together in #169. But since some of the tests still fail locally, I first need to find the issue with that. Could be Mosquitto related, but I'm not sure about it yet.

**Comment by reinweb** [comment]
I am not an expert, but I try to supply my findings:
Here is the same message. First Retained, Second in realtime (not retained).
With **ngrep**  you can see that before the topic (playground/pub)  the retained message starts with `"1....playground"` and the realtime message starts with `"0...playground"`



`ngrep -W byline -T host <broker> -d any`

```
T +0.048657 <broker>:1883 -> <subscriber>:56146 [AP] #11
1...playground/pub20231125_074518
##
T +10.971536 <broker>:1883 -> <subscriber>:56146 [AP] #13
0...playground/pub20231125_074610
```

I have generated my messages with **mosquitto_sub** and **mosquitto_pub** by using this commands

`mosquitto_pub  -u xxx -P xxx -h <broker>  -t playground/pub -m "$(date +%Y%m%d_%H%M%S)" -q 0 --retain --will-payload "myLastWill" --will-qos 0 --will-topic "playground/lastWill" --will-retain  -k 5
`

`mosquitto_sub  -u xxx -P xxx -t 'playground/pub' -k 20 -h <broker>  --verbose  -F '%j' -d --nodelay -q 0 --pretty --retain-as-published
`

the **mosquitto_sub** showed the results below (see the line **"retain:"**:)

```
Client (null) received PUBLISH (d0, q0, r1, m0, 'playground/pub', ... (15 bytes))
{
	"tst":	"2023-11-25T07:45:59.175501Z+0100",
	"topic":	"playground/pub",
	"qos":	0,
	"retain":	1,
	"payloadlen":	15,
	"payload":	"20231125_074518"
}
Client (null) received PUBLISH (d0, q0, r0, m0, 'playground/pub', ... (15 bytes))
{
	"tst":	"2023-11-25T07:46:10.147140Z+0100",
	"topic":	"playground/pub",
	"qos":	0,
	"retain":	0,
	"payloadlen":	15,
	"payload":	"20231125_074610"
}
```

in the repository [https://github.com/bluerhinos/phpMQTT/blob/master/phpMQTT.php](https://github.com/bluerhinos/phpMQTT/blob/master/phpMQTT.php) I was able to extract the retain-flag  in line 518 but I prefer to use your project - so I will be very happy if you add the retain-flag in your project. thx a lot

**Comment by Namoshek** [comment]
The required changes for this have been released as `v2.0.0`. The new version also requires a minimum PHP version of 8.0 as PHP 7 is completely end of life.
